### ISSUE TITLE
error in cl manual examples

### ISSUE THREAD

**Issue by gilFuser** [jamshark70/ddwChucklib-livecode#1]
Hi,
I'm having the first contact with cll now, And I must say it looks really interesting and promising.
I'm getting an error when trying to run a specific example in the manual.

// In this example everything is working fine...

(
BP(\y).free;
PR(\abstractLiveCode).chuck(BP(\y), nil, (
	event: (eventKey: \default),
	defaultParm: \degree,
	parmMap: (degree: (isPitch: true))
));
)

TempoClock.tempo = 140/60;

/y = "12 4| 5 6| 12 |45";

/y+;

/y = "\seq("** *| * *| ** |**", "12456", "*")";

// ... until here. The ones that come below produce the error.

/y = "\ins(\seq("** *| * *| ** |**", "12456", "*"), "*", 7, 0.25)";

/y = "\seq(\ins(\seq("** *| * *| ** |**", "12456", "*"), "*", 7, 0.25), "6,214", "*")";

/y = "\seq("** *| * *| ** |**", "12456", "*")::\ins(, "*", 7, 0.25)::\seq(, "6,214", "*")";

/y-;

// further below this works:

// 1. \rand starts on the downbeat and occupies the whole bar.
/y = "\rand(\ins("1,", "*", 3, 0.5), "13467", "*")";

/y+;

// but from now on I get the error again.

// 2. \rand starts on beat 2
/y = "1,|\rand(\ins("6,", "*", 3, 0.5), "13467", "*")||";

// 3. \rand starts on the 2nd 16th-note of beat 2
/y = "1,|6,\rand(\ins("", "*", 3, 0.5), "13467", "*")  ||";

// 4. \rand starts on the 2nd 16th-note of beat 2
// and stops on the 'and' of 4
/y = "1,|6,\rand(\ins("", "*", 3, 0.5), "13467", "*")  || x";

/y-;

I'm using SC3.9dev in Linux.

best regards,
Gil

ERROR: FATAL ERROR:
Binary op '==' on SequenceNote failed.

PROTECTED CALL STACK:
	Object:numPerformBinaryOpOnNumber	0x55ebf01e0d40
		arg this = *
		arg selector = ==
	a FunctionDef	0x55ebef4e8040
		sourceCode = "<an open Function>"
		arg item2 = *
	ArrayedCollection:do	0x55ebef5df340
		arg this = [ * ]
		arg function = a Function
		var i = 0
	Collection:includesEqual	0x55ebef4e7d40
		arg this = [ * ]
		arg item1 = [ 0, nil, 0.9, nil ]
	a FunctionDef	0x55ebf0feb948
		sourceCode = "{ |item, i|
				// includesEqual: SequenceNotes compare pitch only! Use pitch for matching
				if(~replaceItems.includesEqual(item.item)) {
					// some replacers might need access to the state
					~currentItem = item;
					~currentIndex = i;
					// copy is critical here!
					~items[i] = item.copy.put(\\item, ~seqStream.next(item));
				};
			}"
		arg item = ( 'time': 0, 'item': a SequenceNote, 'dur': 0.25 )
		arg i = 0
	ArrayedCollection:do	0x55ebef5df340
		arg this = [ ( 'time': 0, 'item': a SequenceNote, 'dur': 0.25 ), ( 'time': 0.25, 'item': a SequenceNote, 'dur': 0.5 ), ( 'item': *, 'time': 0.5 ), ( 'time': 0.75, 'item': a SequenceNote, 'dur': 0.5 ), ( 'item': *, 'time': 1 ), ( 'time': 1.25, 'item': a SequenceNote, 'dur': 0.5 ), ( 'item': *, 'time': 1.5 ), ( 'time': 1.75, 'item': a SequenceNote, 'dur': 0.5 ), ( 'item': *, 'time': 2 ), ( 'time': 2.25, 'item': a SequenceNote, 'dur': 0.25 ), ( 'time': 2.5, 'item': a SequenceNote, 'dur': 0.5 ), ( 'item': *, 'time': 2.75 ...etc...
		arg function = a Function
		var i = 0
	a FunctionDef	0x55ebf15c3228
		sourceCode = "{
			if(~doReset or: { ~seqStream.isNil }) {
				~makeStream.();
			};
			~items = ~getUpstreamItems.();
			~modifySource.();
			~items
		}"
	a FunctionDef	0x55ebef109240
		sourceCode = "<an open Function>"
	a FunctionDef	0x55ebef107280
		sourceCode = "<an open Function>"
	Function:prTry	0x55ebee4f2c00
		arg this = a Function
		var result = nil
		var thread = a Routine
		var next = a Function
		var wasInProtectedFunc = true
	Function:protect	0x55ebee4f2180
		arg this = a Function
		arg handler = a Function
		var result = nil
	Proto:use	0x55ebef106d00
		arg this = a Proto
		arg func = a Function
		arg failSilentlyAfterFree = false
		var result = nil
		var saveEnvir = a Proto
	Proto:doesNotUnderstand	0x55ebef108b00
		arg this = a Proto
		arg selector = process
		arg args = nil
		var result = [  ]
		var item = nil
	a FunctionDef	0x55ebf13fedd8
		sourceCode = "<an open Function>"
		arg gen = a Proto
	ArrayedCollection:do	0x55ebef5df340
		arg this = [ a Proto, a Proto, a Proto ]
		arg function = a Function
		var i = 2
	a FunctionDef	0x55ebf13f0248
		sourceCode = "{
			var items;
			// ~baseItems is nil if chain is the outermost generator
			// but if chain is inside a fork, the fork will provide baseItems
			if(~baseItems.notNil) { items = ~getUpstreamItems.() };
			~args.do { |gen|
				if(items.notNil) {
					// this is corrupting the initial state,
					// but this generator will only ever be processed here
					gen.baseItems = items;
				};
				items = gen.process;
			};
			~items = items;
		}"
		var items = [ ( 'time': 0, 'item': a SequenceNote, 'dur': 0.25 ), ( 'time': 0.25, 'item': a SequenceNote, 'dur': 0.5 ), ( 'item': *, 'time': 0.5 ), ( 'time': 0.75, 'item': a SequenceNote, 'dur': 0.5 ), ( 'item': *, 'time': 1 ), ( 'time': 1.25, 'item': a SequenceNote, 'dur': 0.5 ), ( 'item': *, 'time': 1.5 ), ( 'time': 1.75, 'item': a SequenceNote, 'dur': 0.5 ), ( 'item': *, 'time': 2 ), ( 'time': 2.25, 'item': a SequenceNote, 'dur': 0.25 ), ( 'time': 2.5, 'item': a SequenceNote, 'dur': 0.5 ), ( 'item': *, 'time': 2.75 ...etc...
	a FunctionDef	0x55ebef1028c0
		sourceCode = "<an open Function>"
	a FunctionDef	0x55ebef107280
		sourceCode = "<an open Function>"
	Function:prTry	0x55ebee4f2c00
		arg this = a Function
		var result = nil
		var thread = a Routine
		var next = a Function
		var wasInProtectedFunc = true
	Function:protect	0x55ebee4f2180
		arg this = a Function
		arg handler = a Function
		var result = nil
	Proto:use	0x55ebef106d00
		arg this = a Proto
		arg func = a Function
		arg failSilentlyAfterFree = false
		var result = nil
		var saveEnvir = a Proto
	Proto:value	0x55ebef102440
		arg this = a Proto
		arg args = nil
		var result = [  ]
	a FunctionDef	0x55ebef4eabc0
		sourceCode = "<an open Function>"
		arg elem = a Proto
		arg i = 0
	ArrayedCollection:do	0x55ebef5df340
		arg this = [ a Proto ]
		arg function = a Function
		var i = 0
	Collection:collectAs	0x55ebef4ea800
		arg this = [ a Proto ]
		arg function = a Function
		arg class = Array
		var res = [  ]
	a FunctionDef	0x55ebf0ac18c8
		sourceCode = "{ |key(\\baseItems)|
			var out;
			// \">> % getUpstreamItems\".format(~protoID).debug;
			out = key.envirGet.asArray.collect { |item| item.value }.flatten(1);
			// \"<< % getUpstreamItems\".format(~protoID).debug;
			// out.postcs
		}"
		arg key = baseItems
		var out = nil
	a FunctionDef	0x55ebf1b2d6b8
		sourceCode = "{
			~items = ~getUpstreamItems.();
		}"
	a FunctionDef	0x55ebf1b07a58
		sourceCode = "{
			if(~repository.isNil) {
				if(BP.exists(~bpKey)) {
					if(BP(~bpKey)[\\genRepository].notNil) {
						~repository = BP(~bpKey)[\\genRepository];
					} {
						~repository = IdentityDictionary.new;
						BP(~bpKey)[\\genRepository] = ~repository;
					};
				};
			};
			~process.();
			if(~items.size > 0) {
				if(~repository[~parm].isNil) {
					// this happens only the first time the phrase is invoked in a new BP:
					// if nothing on the downbeat, S&H the first real item
					~rep...etc..."
	a FunctionDef	0x55ebf0fee748
		sourceCode = "{ |bp|  // the BP() object -- asPattern should be called @ top level only
			var pat, rest, didProcess = false;
			if(~items.isNil) {
				~items = ~doProcess.(bp);
				~updateDeltas.();
				// \"clGen items: \".post; ~items.postcs;
				didProcess = true;
			};
			// if(~items[0].time > 0) {
			// 	rest = [(time: 0, dur: ~items[0].time, item: Rest(~items[0].item))];
			// };
			// \"rest ++\" also works if 'rest' is nil
			pat = Pseq(rest ++ ~items, 1);
			if(BP.exists(~bpKey) and: { ~parm !=...etc..."
		arg bp = BP('y')
		var pat = nil
		var rest = nil
		var didProcess = false
	a FunctionDef	0x55ebf056d918
		sourceCode = "{ |inval|
			var dur = ~dur.(),
			startTime = thisThread.clock.beats,
			next, pitch, bp, rout;
			if(BP.exists(~bpKey)) {  // this also avoids 'nil'
				bp = BP(~bpKey);
			};

			~asPattern.(bp).embedInStream(inval);
		}"
		arg inval = ( 'collIndex': y, 'timingOffset': 0, 'propagateDownward': a Function, 'eventKey': default, 
  'clock': a TempoClock )
		var dur = 4
		var startTime = 2964
		var next = nil
		var pitch = nil
		var bp = BP('y')
		var rout = nil
	a FunctionDef	0x55ebef105680
		sourceCode = "<an open Function>"
	a FunctionDef	0x55ebef107280
		sourceCode = "<an open Function>"
	Function:prTry	0x55ebee4f2c00
		arg this = a Function
		var result = nil
		var thread = a Routine
		var next = nil
		var wasInProtectedFunc = true
	
CALL STACK:
	Exception:reportError   0x55ebf1a78dd8
		arg this = <instance of Error>
	Nil:handleError   0x55ebf03fda48
		arg this = nil
		arg error = <instance of Error>
	Thread:handleError   0x55ebed7b1828
		arg this = <instance of Thread>
		arg error = <instance of Error>
	Thread:handleError   0x55ebf04adab8
		arg this = <instance of Routine>
		arg error = <instance of Error>
	Thread:handleError   0x55ebf05f8b28
		arg this = <instance of Routine>
		arg error = <instance of Error>
	Thread:handleError   0x55ebf22429f8
		arg this = <instance of Routine>
		arg error = <instance of Error>
	Thread:handleError   0x55ebf0c734a8
		arg this = <instance of Routine>
		arg error = <instance of Error>
	Thread:handleError   0x55ebf1127788
		arg this = <instance of Routine>
		arg error = <instance of Error>
	Thread:handleError   0x55ebf097fb18
		arg this = <instance of Routine>
		arg error = <instance of Error>
	Object:throw   0x55ebf1071988
		arg this = <instance of Error>
	Function:protect   0x55ebed59aa08
		arg this = <instance of Function>
		arg handler = <instance of Function>
		var result = <instance of Error>
	Proto:use   0x55ebf0d3c778
		arg this = <instance of Proto>
		arg func = <instance of Function>
		arg failSilentlyAfterFree = false
		var result = nil
		var saveEnvir = <instance of Proto>
	PatternProxy:embedInStream   0x55ebf06c13e8
		arg this = <instance of PatternProxy>
		arg inval = <instance of Event>
		arg embed = false
		arg default = nil
		var outval = nil
		var count = 0
		var pat = <instance of Pcollect>
		var test = true
		var resetTest = nil
		var stream = <instance of FuncStream>
	Routine:prStart   0x55ebf0b41c18
		arg this = <instance of Routine>
		arg inval = <instance of Event>
^^ The preceding error dump is for ERROR: FATAL ERROR:
Binary op '==' on SequenceNote failed.

**Comment by jamshark70** [comment]
Thanks for the report.

The error is coming from the ddwCommon quark. I had solved this locally, but I hadn't published the change.

I just did that, so, update ddwCommon and the error should go away. https://github.com/jamshark70/ddwCommon/commit/d302e23700905be66db7f5c81fdbe8051f7efb59

I'll keep the issue for now until you confirm the facts -- feel free to close it after that.

BTW I'm about to open a new issue. I'm working on some changes in the way ties are handled for pitch parameters. Other things should stay pretty much the same.
